daml-lf/integration-test-lib/src/main/com/daml/lf/CantonRunner.scala

```diff
@@ -0,0 +1,189 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package integrationtest
+
+import com.daml.bazeltools.BazelRunfiles._
+import com.daml.grpc.adapter.ExecutionSequencerFactory
+import com.daml.jwt.JwtSigner
+import com.daml.jwt.domain.DecodedJwt
+import com.daml.lf.data.Ref
+import com.daml.ledger.api.auth
+import com.daml.ledger.resources.{ResourceContext, ResourceOwner, Resource}
+import com.daml.platform.services.time.TimeProviderType
+import com.daml.ports.{Port, LockedFreePort}
+import com.daml.scalautil.Statement.discard
+import com.daml.timer.RetryStrategy
+import com.google.protobuf.ByteString
+import spray.json.JsString
+
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.Future
+import scala.sys.process.Process
+import java.nio.charset.StandardCharsets
+import java.nio.file.{Path, Files}
+
+object CantonRunner {
+
+  private val logger = org.slf4j.LoggerFactory.getLogger(getClass)
+
+  private[integrationtest] def toJson(s: String): String = JsString(s).toString()
+  private[integrationtest] def toJson(path: Path): String = toJson(path.toString)
+
+  def run(config: CantonConfig, tmpDir: Path)(implicit
+      esf: ExecutionSequencerFactory
+  ): ResourceOwner[Vector[Port]] =
+    new ResourceOwner[Vector[Port]] {
+      import config._
+      protected val cantonConfigFile = tmpDir.resolve("participant.config")
+      protected val cantonLogFile = tmpDir.resolve("canton.log")
+      protected val portFile = tmpDir.resolve("portfile")
+      def info(s: String) = if (debug) logger.info(s)
+
+      override def acquire()(implicit context: ResourceContext): Resource[Vector[Port]] = {
+        def start(): Future[(Vector[Port], Process)] = {
+          val ports =
+            Vector.fill(nParticipants)(LockedFreePort.find() -> LockedFreePort.find())
+          val ledgerPorts = ports.map(_._2.port)
+          val domainPublicApi = LockedFreePort.find()
+          val domainAdminApi = LockedFreePort.find()
```

Comment:
Do you ever free these ports?

Reply:
good catch.

Reply:
In b7d5f01cc90ae734903ccf6b9c95d9da348cfded I unlock the port as soon canton has started.

Reply:
Would it not be better to release when you close canton?
Freeport won't take the port, but it will try it and delay since there will be no lock file.
Also I see you free up `ports`, but are `domainPublicApi` and `domainAdminApi` in ports?

Reply:
5fbaa28c1b6365781c92aa3959d0d0dd00e26cd1